src/lib/mixer_module/mixer_module.hpp

```diff
@@ -0,0 +1,232 @@
+	uint16_t &failsafeValue(int index) { return _failsafe_value[index]; }
+	/** Disarmed values: if ramp_up is true, then disarmedValue < minValue needs to hold */
+	uint16_t &disarmedValue(int index) { return _disarmed_value[index]; }
+	uint16_t &minValue(int index) { return _min_value[index]; }
+	uint16_t &maxValue(int index) { return _max_value[index]; }
+
+protected:
+	void updateParams() override;
+
+private:
+	void handleCommands();
+
+	bool armNoThrottle() const
+	{
+		return (_armed.prearmed && !_armed.armed) || _armed.in_esc_calibration_mode;
+	}
+	static int controlCallback(uintptr_t handle, uint8_t control_group, uint8_t control_index, float &input);
+
+	enum class MotorOrdering : int32_t {
+		PX4 = 0,
+		Betaflight = 1
+	};
+
+	struct Command {
+		enum class Type : int {
+			None,
+			resetMixer,
+			loadMixer
+		};
+		px4::atomic<Type> command{Type::None};
+		const char *mixer_buf;
+		unsigned mixer_buf_length;
+		int result;
+	};
+	Command _command; ///< incoming commands (from another thread)
+
+	/**
+	 * Reorder PWM outputs according to _param_mot_ordering
+	 * @param values PWM values to reorder
+	 */
+	inline void reorderOutputs(uint16_t values[MAX_ACTUATORS]);
+
+	void lock() { do {} while (px4_sem_wait(&_lock) != 0); }
+	void unlock() { px4_sem_post(&_lock); }
+
+	px4_sem_t _lock; /**< lock to protect access to work queue changes (includes ScheduleNow calls from another thread) */
+
+	uint16_t _failsafe_value[MAX_ACTUATORS] {};
+	uint16_t _disarmed_value[MAX_ACTUATORS] {};
+	uint16_t _min_value[MAX_ACTUATORS] {};
+	uint16_t _max_value[MAX_ACTUATORS] {};
+	uint16_t _reverse_output_mask{0}; ///< reverses the interval [min, max] -> [max, min], NOT motor direction
+	pwm_limit_t _pwm_limit;
+
+	uORB::Subscription _armed_sub{ORB_ID(actuator_armed)};
+	uORB::Subscription _safety_sub{ORB_ID(safety)};
+	uORB::SubscriptionCallbackWorkItem _control_subs[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS];
+
+	uORB::PublicationMulti<actuator_outputs_s> _outputs_pub{ORB_ID(actuator_outputs), ORB_PRIO_DEFAULT};
+	uORB::PublicationMulti<multirotor_motor_limits_s> _to_mixer_status{ORB_ID(multirotor_motor_limits), ORB_PRIO_DEFAULT}; 	///< mixer status flags
+
+	actuator_controls_s _controls[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS] {};
+	actuator_armed_s _armed{};
+
+	hrt_abstime _time_last_mix{0};
+	unsigned _max_topic_update_interval_us{0}; ///< max _control_subs topic update interval (0=unlimited)
+
+	bool _safety_off{false}; ///< State of the safety button from the subscribed _safety_sub topic
+	bool _throttle_armed{false};
+
+	MixerGroup *_mixers{nullptr};
+	uint32_t _groups_required{0};
+	uint32_t _groups_subscribed{0};
+
+	const bool _support_esc_calibration;
+
+	bool _wq_switched{false};
+
+	OutputModuleInterface &_interface;
+
+	perf_counter_t _control_latency_perf;
+
+	DEFINE_PARAMETERS(
+		(ParamInt<px4::params::MC_AIRMODE>) _param_mc_airmode,   ///< multicopter air-mode
```

Comment:
We should think of a different way to handle this eventually. Seeing the "Multicopter Attitude Control" param group on a plane deeply bothers me.

Reply:
Agreed. Maybe just move the param to a mixer group?